### PR TITLE
[#51, #323, #426, #479, #450] Query caching API, Method renamings, fetch strategy support for non-correlations, API for joins after correlation and automatic subtype discovery for updatable attributes

### PR DESCRIPTION
**Pull request by beikov**
## Related Issue

#51 #323 #426 #497 #450

### CODE REVIEW

File: documentation/src/main/asciidoc/core/manual/en_US/05_predicate_builder.adoc [hunk 1]
```diff
@@ -20,7 +20,7 @@ There are multiple different entry methods to cover all possible usage scenarios
 e.g. in the `WHERE` clause the entry methods are named `where()`, `whereExists()` etc.
 The following list of possible entry methods refers to `WHERE` clause entry methods for easier readability.
 
-link:{core_jdoc}/persistence/WhereBuilder.html#whereExpression(java.lang.String)[`where(String expression)`]::
+link:{core_jdoc}/persistence/WhereBuilder.html#where(java.lang.String)[`where(String expression)`]::
```
Comment (Mobe91) [inline documentation/src/main/asciidoc/core/manual/en_US/05_predicate_builder.adoc]: Shouldn't this be setWhereExpression?

File: documentation/src/main/asciidoc/core/manual/en_US/12_query_api.adoc [hunk 2]
```diff
@@ -71,6 +71,12 @@ This allows for disabling optimizations when encountering bugs or fine tuning on
 If you encounter, that you make use of a configuration very often, you should consider configuring the property globally via the link:{core_jdoc}/persistence/spi/CriteriaBuilderConfiguration.html[`CriteriaBuilderConfiguration`]
 and only switch to a different value when needed.
 
+=== Query result caching
+
+Query result caching can be enabled by invoking the link:{core_jdoc}/persistence/CommonQueryBuilder.html#setCacheable(boolean)[`setCacheable(boolean)`] method on a query builder.
+
+Note that additional configuration might be required for the caching to work properly. For details, refer to the documentation of you JPA provider.
```
Comment (Mobe91) [inline documentation/src/main/asciidoc/core/manual/en_US/12_query_api.adoc]: of *your* JPA provider

File: documentation/src/main/asciidoc/entity-view/manual/en_US/05_fetch_strategies.adoc [hunk 3]
```diff
@@ -1,14 +1,12 @@
 [[anchor-fetch-strategies]]
 == Fetch strategies
 
-There are multiple different fetch strategies available for fetching. Right now, a fetch strategy can only be applied on a correlated mapping,
-but soon will also be able to be specified for any other mapping. Also see https://github.com/Blazebit/blaze-persistence/issues/323[#323] for further information.
+There are multiple different fetch strategies available for fetching. A fetch strategy can applied on all kinds mappings,
```
Comment (Mobe91) [inline documentation/src/main/asciidoc/entity-view/manual/en_US/05_fetch_strategies.adoc]: A fetch strategy can *be* applied *to* all kinds *of* mappings

File: documentation/src/main/asciidoc/entity-view/manual/en_US/05_fetch_strategies.adoc [hunk 4]
```diff
@@ -1,14 +1,12 @@
...
+except for `@MappingParameter` and `@MappingSubquery`. These mappings will always use a `JOIN` strategy, i.e. the mapping will put into the main query.
```
Comment (Mobe91) [inline documentation/src/main/asciidoc/entity-view/manual/en_US/05_fetch_strategies.adoc]: the mapping will *be* put into the main query

File: documentation/src/main/asciidoc/entity-view/manual/en_US/05_fetch_strategies.adoc [hunk 5]
```diff
@@ -1,14 +1,12 @@
...
 
-NOTE: `@MappingSubquery` is the only mapping that doesn't support overriding the fetch strategy. It will always use the `JOIN` strategy, i.e. it will put the subquery into the main query.
+Any attribute in an entity view can be fetched separately by specifying a fetch strategy other tha `JOIN`.
```
Comment (Mobe91) [inline documentation/src/main/asciidoc/entity-view/manual/en_US/05_fetch_strategies.adoc]: other *than*

File: documentation/src/main/asciidoc/entity-view/manual/en_US/05_fetch_strategies.adoc [hunk 6]
```diff
@@ -1,14 +1,12 @@
...
 
-// Any attribute in an entity view can be fetched separately(that's the idea, but only implemented for correlation right now)
-
-Every fetch strategy has some pros and cons, but generally, the `JOIN` fetch strategy is the one that is good enough most of the time.
+The fetch strategies all have some pros and cons, but generally, the `JOIN` fetch strategy is the one that is good enough most of the time.
```
Comment (Mobe91) [inline documentation/src/main/asciidoc/entity-view/manual/en_US/05_fetch_strategies.adoc]: Every fetch strategy has pros and cons but most of the time, the `JOIN` fetch strategy is a good choice.

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/metamodel/MetamodelBuildingContextImpl.java [hunk 7]
```diff
@@ -421,6 +421,25 @@ public int compare(Class<?> o1, Class<?> o2) {
...
+            }
+        });
+        for (Class<?> clazz : viewMappings.keySet()) {
+            if (clazz.isAssignableFrom(entityViewClass) && entityViewClass != clazz) {
+                supertypes.add(clazz);
+            }
+        }
+
+        return supertypes;
+
```
Comment (Mobe91) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/metamodel/MetamodelBuildingContextImpl.java]: remove newline

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/objectbuilder/ViewTypeObjectBuilderTemplate.java [hunk 8]
```diff
@@ -711,29 +713,27 @@ private void applyCorrelatedSubviewMapping(CorrelatedAttribute<? super T, ?> cor
...
-            String correlationKeyExpression = mapperBuilder.getMapping(AbstractAttribute.stripThisFromMapping(correlatedAttribute.getCorrelationBasis()));
+            Class<?> correlationBasisType = getCorrelationBasisType(AbstractAttribute.stripThisFromMapping(correlationBasis));
+            Class<?> correlationBasisEntity = getCorrelationBasisEntityType(AbstractAttribute.stripThisFromMapping(correlationBasis), correlationBasisType);
+            String correlationKeyExpression = mapperBuilder.getMapping(AbstractAttribute.stripThisFromMapping(correlationBasis));
 
-            mapperBuilder.addMapper(createMapper(correlationKeyExpression, subviewAliasPrefix, correlatedAttribute.getFetches()));
+            mapperBuilder.addMapper(createMapper(correlationKeyExpression, subviewAliasPrefix, attribute.getFetches()));
 
-            if (!correlatedAttribute.isCollection()) {
+            if (!attribute.isCollection()) {
```
Comment (Mobe91) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/objectbuilder/ViewTypeObjectBuilderTemplate.java]: negative logic

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/objectbuilder/ViewTypeObjectBuilderTemplate.java [hunk 9]
```diff
@@ -669,28 +673,26 @@ private void applyCorrelatedSubviewMapping(CorrelatedAttribute<? super T, ?> cor
...
 
-            mapperBuilder.addMapper(createMapper(mapperBuilder.getMapping(AbstractAttribute.stripThisFromMapping(correlatedAttribute.getCorrelationBasis()), correlationBasisEntity), subviewAliasPrefix, correlatedAttribute.getFetches()));
+            mapperBuilder.addMapper(createMapper(mapperBuilder.getMapping(AbstractAttribute.stripThisFromMapping(correlationBasis), correlationBasisEntity), subviewAliasPrefix, attribute.getFetches()));
 
             if (batchSize == -1) {
                 batchSize = 1;
             }
 
-            if (!correlatedAttribute.isCollection()) {
+            if (!attribute.isCollection()) {
```
Comment (Mobe91) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/objectbuilder/ViewTypeObjectBuilderTemplate.java]: negative logic

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/objectbuilder/ViewTypeObjectBuilderTemplate.java [hunk 10]
```diff
@@ -827,49 +827,46 @@ private void applySubqueryMapping(SubqueryAttribute<?, ?> attribute, TupleElemen
...
-            CorrelationProviderFactory factory = CorrelationProviderHelper.getFactory(correlatedAttribute.getCorrelationProvider());
+            CorrelationProviderFactory factory = CorrelationProviderHelper.getFactory(correlationProvider);
 
             if (batchSize == -1) {
                 batchSize = 1;
             }
 
-            String basicAttributePath = getAttributePath(attributePath, correlatedAttribute, false);
-            if (!correlatedAttribute.isCollection()) {
+            if (!attribute.isCollection()) {
```
Comment (Mobe91) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/objectbuilder/ViewTypeObjectBuilderTemplate.java]: negative logic

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/objectbuilder/ViewTypeObjectBuilderTemplate.java [hunk 11]
```diff
@@ -890,31 +887,29 @@ private void applyBasicCorrelatedMapping(CorrelatedAttribute<?, ?> correlatedAtt
...
+            String correlationKeyExpression = mapperBuilder.getMapping(AbstractAttribute.stripThisFromMapping(correlationBasis));
 
-            mapperBuilder.addMapper(createMapper(correlationKeyExpression, subviewAliasPrefix, correlatedAttribute.getFetches()));
+            mapperBuilder.addMapper(createMapper(correlationKeyExpression, subviewAliasPrefix, attribute.getFetches()));
 
-            CorrelationProviderFactory factory = CorrelationProviderHelper.getFactory(correlatedAttribute.getCorrelationProvider());
+            CorrelationProviderFactory factory = CorrelationProviderHelper.getFactory(correlationProvider);
 
-            if (!correlatedAttribute.isCollection()) {
+            if (!attribute.isCollection()) {
```
Comment (Mobe91) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/objectbuilder/ViewTypeObjectBuilderTemplate.java]: negative logic

File: documentation/src/main/asciidoc/core/manual/en_US/05_predicate_builder.adoc (hunk 1 above)
Comment (beikov) [reply documentation/src/main/asciidoc/core/manual/en_US/05_predicate_builder.adoc]: Nope, that's on purpose. This is the entry point for starting with the predicate builder API. I just wrongly referenced a different method there before

**Pull request merged**